### ISSUE TITLE
core-dns addon pod is stuck in pending state

### ISSUE THREAD

**Issue by soumyalj** [rancher/rke#1944]
**RKE version:** rke version v1.1.0-rc10

Steps:
1. Create a RKE cluster using K8s version v1.16.7-rancher1-2
Cluster config: 1 etcd 2 control node and 2 workers
2. Upgrade the cluster using K8s version v1.17.3-rancher1-2

One of the core-dns pod is stuck in pending state as below :

```
soumyas-MBP:rke soumya$ kubectl get pod --kubeconfig kube_config_clusterzeroworker.yml -l k8s-app=kube-dns -n kube-system 
NAME                       READY   STATUS    RESTARTS   AGE
coredns-5c59fd465f-lsp2b   1/1     Running   0          16m
coredns-7c5566588d-l6hh7   0/1     Pending   0          16m
coredns-7c5566588d-vxxbv   1/1     Running   0          15m
```
kubectl describe of the pending pod `coredns-7c5566588d-l6hh7`  has the below error 
```
Events:
  Type     Reason            Age        From               Message
  ----     ------            ----       ----               -------
  Warning  FailedScheduling  <unknown>  default-scheduler  0/5 nodes are available: 2 node(s) didn't match pod affinity/anti-affinity, 3 node(s) didn't match node selector.
  Warning  FailedScheduling  <unknown>  default-scheduler  0/5 nodes are available: 2 node(s) didn't match pod affinity/anti-affinity, 3 node(s) didn't match node selector.

```

Note: This issue does not happen in a cluster with 3 worker nodes

**Comment by deniseschannon** [comment]
These errors aren't related to the RKE upgrade logic changes, but to the scheduling logic in the coredns template.

**Comment by Oats87** [comment]
This appears to happen due to the logic we have for the autoscaler + coredns deployment.

https://github.com/rancher/kontainer-driver-metadata/blob/dev-v2.4/rke/templates/coredns.go#L275 indicates that `nodesPerReplica` is `4`, and according to the `cluster-proportional-autoscaler` logic of `ceil( nodes * 1/nodesPerReplica )` and `5` nodes in the cluster, we would end at a desired number of replicas as 2.

In conjunction with the `RollingUpdate` strategy being set with a `MaxUnavailable` of `1`, the third replica cannot be scheduled.

More investigation will be done into this. @soumyalj did this not occur in earlier RKE versions for CoreDNS upgrades?

**Comment by deniseschannon** [comment]
Actually, the new add-on strategy was added by @mrajashree so it could have been triggered by that. Re-assigning to @mrajashree to look into.

**Comment by mrajashree** [comment]
@soumyalj Fix is available with v1.1.0-rc13

**Comment by soumyalj** [comment]
Validated with rke version v1.1.0-rc13 
Steps:
1. Create a RKE cluster using K8s version v1.16.7-rancher1-2
2. Cluster config: 1 etcd 2 control node and 2 workers
3. Upgrade the cluster using K8s version v1.17.3-rancher1-2
There ware no Core-dns pods stuck in pending state. 
The coredns pods were in running state as expected
```
soumyas-MBP:rke soumya$ kubectl get pod --kubeconfig kube_config_clusterzeroworker.yml -l k8s-app=kube-dns -n kube-system 
NAME                       READY   STATUS    RESTARTS   AGE
coredns-7c5566588d-c5jt9   1/1     Running   0          58s
coredns-7c5566588d-q8754   1/1     Running   1          105s


```
